blame/blame.go

```diff
@@ -0,0 +1,286 @@
+// Package blame contains blaming functionality for files in the repo.
+//
+// Blaming a file is finding what commit was the last to modify each of
+// the lines in the file, therefore the output of a blaming operation is
+// usualy a slice of commits, one commit per line in the file.
+//
+// This package also provides a pretty print function to output the
+// results of a blame in a similar format to the git-blame command.
+package blame
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"unicode/utf8"
+
+	"gopkg.in/src-d/go-git.v2"
+	"gopkg.in/src-d/go-git.v2/core"
+	"gopkg.in/src-d/go-git.v2/diff"
+	"gopkg.in/src-d/go-git.v2/revlist"
+)
+
+// Blame returns the last commit that modified each line of a file in
+// a repository.
+//
+// The file to blame is identified by the input arguments: repo, commit and path.
+// The output is a slice of commits, one for each line in the file.
+//
+// Blaming a file is a two step process:
+//
+// 1. Create a linear history of the commits affecting a file. We use
+//    revlist.New for that.
+//
+// 2. Then build a graph with a node for every line in every file in
+//    the history of the file.
+//
+//    Each node (line) holds the commit where it was introduced or
+//    last modified. To achieve that we use the FORWARD algorithm
+//    described in Zimmermann, et al. "Mining Version Archives for
+//    Co-changed Lines", in proceedings of the Mining Software
+//    Repositories workshop, Shanghai, May 22-23, 2006.
+//
+//    Each node is asigned a commit: Start by the nodes in the first
+//    commit. Assign that commit as the creator of all its lines.
+//
+//    Then jump to the nodes in the next commit, and calculate the diff
+//    between the two files. Newly created lines get
+//    assigned the new commit as its origin. Modified lines also get
+//    this new commit. Untouched lines retain the old commit.
+//
+//    All this work is done in the assignOrigin function.
+//
+//    This function holds all the internal relevant data in a blame
```

Comment:
Rephrase proposal: `All this work is done in the assignOrigin function which holds all the internal relevant data in a blame struct, that is not exported.`